src/StreamUtils.bsv

```diff
@@ -0,0 +1,380 @@
+endfunction
+
+(* synthesize *)
+module mkStreamConcat (StreamConcat ifc);
+
+    FIFOF#(DataStream) inputFifoA <- mkFIFOF;
+    FIFOF#(DataStream) inputFifoB <- mkFIFOF;
+    FIFOF#(DataStream) outputFifo <- mkFIFOF;
+
+    FIFOF#(StreamWithPtr) prepareFifoA <- mkFIFOF;
+    FIFOF#(StreamWithPtr) prepareFifoB <- mkFIFOF;
+
+    Reg#(DataBytePtr) bytePtrRegA <- mkReg(0);
+    Reg#(DataBytePtr) bytePtrRegB <- mkReg(0);
+    Reg#(DataBytePtr) remainBytePtrReg <- mkReg(0);
+
+    Reg#(Bool) hasRemainReg <- mkReg(False);
+    Reg#(Bool) hasLastRemainReg <- mkReg(False);
+    Reg#(Bool) isStreamAEnd <- mkReg(False);
+
+    Reg#(DataStream) remainStreamReg <- mkRegU;
+    
+
+    rule prepareStreamA;
+        let streamA = inputFifoA.first;
+        inputFifoA.deq;
+        DataBytePtr bytePtr = convertByteEn2BytePtr(streamA.byteEn);
+        prepareFifoA.enq(StreamWithPtr {
+            stream: streamA,
+            bytePtr: bytePtr
+        });
+    endrule
+
+    rule prepareStreamB;
+        let streamB = inputFifoB.first;
+        inputFifoB.deq;
+        DataBytePtr bytePtr = convertByteEn2BytePtr(streamB.byteEn);
+        prepareFifoB.enq(StreamWithPtr {
+            stream: streamB,
+            bytePtr: bytePtr
+        });
+    endrule
+
+    rule concatStream;
+        // Only the remain data
+        if (hasRemainReg && hasLastRemainReg) begin
+            outputFifo.enq(remainStreamReg);
+            hasRemainReg <= False;
+            isStreamAEnd <= False;
+        end
+        // StreamB or streamB + the remain data
+        else if (prepareFifoB.notEmpty && isStreamAEnd) begin
+            let streamB = prepareFifoB.first.stream;
+            let bytePtrB = prepareFifoB.first.bytePtr;
+            prepareFifoB.deq;
+            streamB.isFirst = False;
+            if (hasRemainReg) begin
+                let {concatStream, remainStream, remainBytePtr} = getConcatStream(remainStreamReg, streamB, remainBytePtrReg, bytePtrB);
+                hasRemainReg     <= !isByteEnZero(remainStream.byteEn);
+                hasLastRemainReg <= streamB.isLast;
+                remainStreamReg  <= remainStream;
+                remainBytePtrReg <= remainBytePtr;
+                outputFifo.enq(concatStream);
+            end
+            else begin
+                outputFifo.enq(streamB);
+            end
+            isStreamAEnd <= !streamB.isLast;
+        end
+        // StreamA or StreamA + first StreamB
+        else if (prepareFifoA.notEmpty) begin
+            let streamA = prepareFifoA.first.stream;
+            let bytePtrA = prepareFifoA.first.bytePtr;
+            // Only StreamA frame
+            if (!streamA.isLast) begin
+                outputFifo.enq(streamA);
+                prepareFifoA.deq;
+                isStreamAEnd <= False;
+            end 
+            // the last StreamA + the first StreamB
+            else if (streamA.isLast && prepareFifoB.notEmpty) begin
```

Comment:
infact, the condition in this else if statement is not necessary. the bsv has implicit conditions, so if the prepareFifoB is empty, this `else if` branch will not be executed.

but you write the condition here is good for code readability, you can keep the condition here, but this lead to another problem: this if statement has other possiblity that is not covered. you should add a `else` branch, and add some code there, maybe only some comment like "in all other condition, wait for stream B to be ready"

Reply:
Yes, the condition here is not necessary but good for understanding. I have added a `else` branch where just waiting for streamB in this clock.